compiler/modules/nback/module.bal

```diff
@@ -56,12 +57,42 @@ public function buildModule(bir:Module birMod, *Options options) returns [llvm:M
         DISubprogram? diFunc = di == () ? () : diFuncs[i];
         Scaffold scaffold = new(mod, llFuncs[i], diFunc, builder, defn, code);
         buildPrologue(builder, scaffold, defn.position);
-        check buildFunctionBody(builder, scaffold, code);
+        bir:Label[] sorted = [];
+        topoSort(code.blocks, 0, sorted);
+        int orphaned = code.blocks.length() - sorted.length();
+        if orphaned != 0 {
+            panic err:impossible(orphaned.toString() + " orphan basic block" + (orphaned == 1 ? "" : "s"));
+        }
+        check buildFunctionBody(builder, scaffold, code.blocks, sorted);
     }
     check birMod.finish();
     return [llMod, createTypeUsage(mod.usedSemTypes)];
 }
 
+function topoSort(bir:BasicBlock[] bbs, bir:Label l, bir:Label[] result) {
+    if result.indexOf(l) != () {
+        return;
+    }
+    bir:BasicBlock bb = bbs[l];
+    bir:Label? onPanic = bb.onPanic;
+    if onPanic != () {
+        topoSort(bbs, onPanic, result);
+    }
+    var insns = bb.insns;
+    var insnsLen = insns.length();
+    if insnsLen > 0 {
+        var insn = insns[insnsLen - 1];
+        if insn is bir:BranchInsn && !insn.backward {
+            topoSort(bbs, insn.dest, result);
+        }
+        else if insn is bir:CondBranchInsn {
+            topoSort(bbs, insn.ifTrue, result);
+            topoSort(bbs, insn.ifFalse, result);
+        }
+    }
+    result.push(bb.label);
```

Comment:
Isn't this `l`?

Is this code really right? Doesn't the test for `result.indexOf(l)` need to be just before the `push`, since the recursive calls to `topoSort` can add members to the result.

Reply:
yes it's `l`, fixed.

Checking in ether place is correct, checking it in the beginning is more efficient. Having a boolean array makes it clearer.